core/src/main/java/com/linecorp/armeria/client/routing/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,154 @@
+     * @param zNode a zNode path eg. /groups/productionGroups
+     * @param sessionTimeout   zookeeper session timeout
+     * @param nodeValueToEndpointList a function to convert zNode value to a List of {@link Endpoint}
+     */
+    public ZookeeperEndpointGroup(String hostPort, String zNode, int sessionTimeout,
+                                  Function<byte[], List<Endpoint>> nodeValueToEndpointList) {
+
+        requireNonNull(hostPort);
+        requireNonNull(zNode);
+        requireNonNull(nodeValueToEndpointList);
+        this.hostPort = hostPort;
+        this.znode = zNode;
+        this.nodeValueToEndpointList = nodeValueToEndpointList;
+        try {
+            zooKeeper = new ZooKeeper(hostPort, sessionTimeout, event -> {
+                if (event.getState() == KeeperState.SyncConnected) {
+                    connectionLatch.countDown();
+                }
+
+            });
+            connectionLatch.await();
+
+            Stat stat = zooKeeper.exists(zNode, this);
+            byte[] nodeData;
+            if (stat != null) {
+                nodeData = zooKeeper.getData(zNode, false, null);
+                endpointList = nodeValueToEndpointList.apply(nodeData);
+            }
+
+        } catch (IOException e) {
+            logger.error("error occurred while connecting zooKeeper " + e.getMessage());
+            e.printStackTrace();
+        } catch (InterruptedException ignored) {
+            //ignored
+        } catch (KeeperException ke) {
+            logger.error("error occurred while reading data from zookeeper ");
+        }
+
+    }
+
+    @Override
+    public List<Endpoint> endpoints() {
+        return this.endpointList;
+
+    }
+
+    @Override
+    public void process(WatchedEvent event) {
+        String path = event.getPath();
+        if (event.getType() != Event.EventType.None) {
+            if (path != null && path.equals(znode)) {
+                // Something has changed on the node, let's find out
+                zooKeeper.exists(znode, true, this, null);
+            }
+        }
+
+    }
+
+    @Override
+    public void processResult(int responseCodeInt, String path, Object ctx, Stat stat) {
+        boolean exists;
+        Code responseCode = Code.get(responseCodeInt);
+        switch (responseCode) {
+            case OK:
+                exists = true;
+                break;
+            case NONODE:
+                exists = false;
+                break;
+            case SESSIONEXPIRED:
+            case NOAUTH:
+                return;
+            default:
+                // Retry errors
+                zooKeeper.exists(znode, true, this, null);
+                return;
+        }
+
+        byte[] nodeData = null;
+        if (exists) {
+            try {
+                nodeData = zooKeeper.getData(znode, false, null);
+            } catch (KeeperException e) {
+                e.printStackTrace();
+            } catch (InterruptedException e) {
+                return;
```

Comment:
I think the exception is better to be ignored ,since it  just remind the user a exception occurs and will not influence business . see my previous comment for the reason. what do you think ?